File: src/main/java/xyz/avarel/waffles/nodes/DefNode.java
package xyz.avarel.waffles.nodes;

import com.oracle.truffle.api.CompilerDirectives;
import com.oracle.truffle.api.dsl.NodeChild;
import com.oracle.truffle.api.dsl.NodeField;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.frame.FrameSlot;
import com.oracle.truffle.api.frame.FrameSlotKind;
import com.oracle.truffle.api.frame.VirtualFrame;

@NodeChild("valueNode")
@NodeField(name = "slot", type = FrameSlot.class)
public abstract class DefNode extends WaffleNode {
    protected abstract FrameSlot getSlot();

    @Specialization(guards = "isIntKind()")
    protected int writeInt(VirtualFrame virtualFrame, int value) {
        virtualFrame.setInt(getSlot(), value);
        return value;
    }

    @Specialization(guards = "isBooleanKind()")
    protected boolean writeBoolean(VirtualFrame virtualFrame, boolean value) {
        virtualFrame.setBoolean(getSlot(), value);
        return value;
    }

    @Specialization(replaces = {"writeInt", "writeBoolean"})
    protected Object writeObject(VirtualFrame virtualFrame, Object value) {
        FrameSlot slot = getSlot();
        if (slot.getKind() != FrameSlotKind.Object) {
            CompilerDirectives.transferToInterpreterAndInvalidate();
            slot.setKind(FrameSlotKind.Object);
        }
        virtualFrame.setObject(slot, value);
        return value;
    }

    protected boolean isIntKind() {
        return getSlot().getKind() == FrameSlotKind.Int;
    }

    protected boolean isBooleanKind() {
        return getSlot().getKind() == FrameSlotKind.Boolean;
    }
}

File: src/main/kotlin/xyz/avarel/waffles/nodes/DefNode.kt
package xyz.avarel.waffles.nodes

import com.oracle.truffle.api.CompilerDirectives
import com.oracle.truffle.api.dsl.NodeChild
import com.oracle.truffle.api.dsl.NodeField
import com.oracle.truffle.api.dsl.Specialization
import com.oracle.truffle.api.frame.FrameSlot
import com.oracle.truffle.api.frame.FrameSlotKind
import com.oracle.truffle.api.frame.VirtualFrame

@NodeChild("valueNode")
@NodeField(name = "slot", type = FrameSlot::class)
abstract class DefNode : WaffleNode() {
    protected abstract val slot: FrameSlot

    protected val isIntKind: Boolean get() = slot.kind == FrameSlotKind.Int
    protected val isBooleanKind: Boolean get() = slot.kind == FrameSlotKind.Boolean

    @Specialization(guards = ["isIntKind()"])
    protected fun writeInt(virtualFrame: VirtualFrame, value: Int): Int {
        virtualFrame.setInt(slot, value)
        return value
    }

    @Specialization(guards = ["isBooleanKind()"])
    protected fun writeBoolean(virtualFrame: VirtualFrame, value: Boolean): Boolean {
        virtualFrame.setBoolean(slot, value)
        return value
    }

    @Specialization(replaces = ["writeInt", "writeBoolean"])
    protected fun writeObject(virtualFrame: VirtualFrame, value: Any): Any {
        val slot = slot
        if (slot.kind != FrameSlotKind.Object) {
            CompilerDirectives.transferToInterpreterAndInvalidate()
            slot.kind = FrameSlotKind.Object
        }
        virtualFrame.setObject(slot, value)
        return value
    }
}

